server/multitenancy/tenant_resolver.ts

```diff
@@ -0,0 +1,137 @@
+ * @param config security plugin config.
+ * @param cookie cookie extracted from the request. The cookie should have been parsed by AuthenticationHandler.
+ * pass it as parameter instead of extracting again.
+ * @param authInfo authentication info, the Elasticsearch authinfo API response.
+ *
+ * @returns user preferred tenant of the request.
+ */
+
+export const PRIVATE_TENANT: string = '__user__';
+export const GLOBAL_TENANT: string = '';
+
+export function resolveTenant(
+  request: KibanaRequest,
+  username: string,
+  availabeTenants: any,
+  config: SecurityPluginConfigType,
+  cookie: SecuritySessionCookie
+): string {
+  let selectedTenant: string = undefined;
+  if (request.query && ((request.query as any).security_tenant || (request.query as any).securitytenant)) {
+    selectedTenant = (request.query as any).security_tenant
+      ? (request.query as any).security_tenant
+      : (request.query as any).securitytenant;
+  } else if (request.headers.securitytenant || request.headers.security_tenant) {
+    selectedTenant = request.headers.securitytenant
+      ? (request.headers.securitytenant as string)
+      : (request.headers.security_tenant as string);
+  } else if (cookie.tentent) {
+    selectedTenant = cookie.tentent;
+  } else {
+    selectedTenant = undefined;
+  }
+
+  const preferredTenants = config.multitenancy.tenants.preferred;
+  const globalTenantEnabled = config.multitenancy.tenants.enable_global;
+  const privateTenantEnabled = config.multitenancy.tenants.enable_private;
+
+  return resolve(username, selectedTenant, preferredTenants, availabeTenants, globalTenantEnabled, privateTenantEnabled);
+}
+
+/**
+ * Determines whether the request requires tenant info.
+ * @param request kibana request.
+ *
+ * @returns true if the request requires tenant info, otherwise false.
+ */
+export function isMultitenantPath(request: KibanaRequest): boolean {
+  return (
+    request.url.path.startsWith('/elasticsearch') ||
+    request.url.path.startsWith('/api') ||
+    request.url.path.startsWith('/app') ||
+    request.url.path === '/'
+  );
+}
+
+function resolve(
+  username: string,
+  requestedTenant: string,
+  preferredTenants: string[],
+  availableTenants: any, // is an object like { tenant_name_1: true, tenant_name_2: false, ... }
+  globalTenantEnabled: boolean,
+  privateTenantEnabled: boolean
+): string {
+  let availableTenantsClone = cloneDeep(availableTenants);
+  delete availableTenantsClone[username];
+
+  if (!globalTenantEnabled && !privateTenantEnabled && isEmpty(availableTenantsClone)) {
+    return undefined;
+  }
+
+  if (requestedTenant) {
+    if (availableTenants[requestedTenant]) {
+      return requestedTenant;
+    }
+
+    if (privateTenantEnabled && availableTenants[username] && requestedTenant === PRIVATE_TENANT /* || requestedTenant === 'private'*/) {
+      return PRIVATE_TENANT;
+    }
+
+    if (globalTenantEnabled && requestedTenant === GLOBAL_TENANT /*|| requestedTenant === 'global'*/) {
```

Comment:
is it comments or unused code?
```javascript
/*|| requestedTenant === 'global'*/
```

Reply:
oh, in the existing plugin implementation, it sometimes use `global` for global tenant, and sometimes empty string `""`, I want to remove it from the new Kibana plugin, not quite sure if this "global" also used in backend ES plugin, so leave a comment here to remind me :P